src/Disks/DiskLocal.cpp

```
@@ -409,67 +458,212 @@ void DiskLocal::applyNewSettings(const Poco::Util::AbstractConfiguration & confi
+    return {};
 }
 
-DiskLocalReservation::~DiskLocalReservation()
+bool DiskLocal::canRead() const noexcept
+try
+{
+    if (FS::canRead(fs::path(disk_path) / disk_checker_path))
+    {
+        auto magic_number = readDiskCheckerMagicNumber();
+        if (magic_number && *magic_number == disk_checker_magic_number)
+            return true;
+    }
+    return false;
+}
+catch (...)
+{
+    LOG_WARNING(logger, "Cannot achieve read over the disk directory: {}", disk_path);
+    return false;
+}
+
+struct DiskWriteCheckData
+{
+    constexpr static size_t PAGE_SIZE = 4096;
+    char data[PAGE_SIZE]{};
+    DiskWriteCheckData()
+    {
+        static const char * magic_string = "ClickHouse disk local write check";
+        static size_t magic_string_len = strlen(magic_string);
+        memcpy(data, magic_string, magic_string_len);
+        memcpy(data + PAGE_SIZE - magic_string_len, magic_string, magic_string_len);
+    }
+};
+
+bool DiskLocal::canWrite() const noexcept
+try
+{
+    static DiskWriteCheckData data;
+    String tmp_template = fs::path(disk_path) / "";
+    {
+        auto buf = WriteBufferFromTemporaryFile::create(tmp_template);
+        buf->write(data.data, data.PAGE_SIZE);
+        buf->sync();
+    }
+    return true;
+}
+catch (...)
+{
+    LOG_WARNING(logger, "Cannot achieve write over the disk directory: {}", disk_path);
+    return false;
+}
+
+bool DiskLocal::setup()
 {
     try
     {
-        std::lock_guard lock(DiskLocal::reservation_mutex);
-        if (disk->reserved_bytes < size)
+        fs::create_directories(disk_path);
+    }
+    catch (...)
+    {
+        LOG_ERROR(logger, "Cannot create the directory of disk {} ({}).", name, disk_path);
+        throw;
+    }
+
+    try
+    {
+        if (!FS::canRead(disk_path))
+            throw Exception(ErrorCodes::PATH_ACCESS_DENIED, "There is no read access to disk {} ({}).", name, disk_path);
```

@amosbird should it `return false` here instead?

Otherwise this method never returns `false` and return value is simply useless, and the line `readonly = !setup()` does not makes a lot of sense.

---

Something like this https://github.com/ClickHouse/ClickHouse/pull/43143/commits/ba6b5bd7605745b1db247646f7f2e5b089516fcf

---

IIRC, it's because `FS::canRead` might also throw exception, so instead of returning false we throw exception instead.

> return value is simply useless, 

Indeed. The function signature should be changed to `void setup()`.

---

https://github.com/ClickHouse/ClickHouse/blob/0b4e643c2764543911b0b4c0c3745846af05742e/src/Common/filesystemHelpers.cpp#L313-L326

Looking at `FS::canRead` and it seems that it makes sense to do something like this:

```c++
try {
  readonly = !FS::canRead();
} catch (...) {
  readonly = true;
}
```

---

But `readonly` is not `broken`.

---

>  @amosbird should it return false here instead?

Ah yes, it should.